File: gcc/testsuite/c-c++-common/goacc/routine-4.c
/* Test invalid use of clauses with routine.  */
/* { dg-do compile } */

#pragma acc routine gang worker /* { dg-error "multiple loop axes" } */
void
f1 (void)
{
}

#pragma acc routine worker gang /* { dg-error "multiple loop axes" } */
void
f1a (void)
{
}

#pragma acc routine gang vector /* { dg-error "multiple loop axes" } */
void
f2 (void)
{
}

#pragma acc routine vector gang /* { dg-error "multiple loop axes" } */
void
f2a (void)
{
}

#pragma acc routine gang seq /* { dg-error "multiple loop axes" } */
void
f3 (void)
{
}

#pragma acc routine seq gang /* { dg-error "multiple loop axes" } */
void
f3a (void)
{
}

#pragma acc routine worker vector /* { dg-error "multiple loop axes" } */
void
f4 (void)
{
}

#pragma acc routine vector worker /* { dg-error "multiple loop axes" } */
void
f4a (void)
{
}

#pragma acc routine worker seq /* { dg-error "multiple loop axes" } */
void
f5 (void)
{
}

#pragma acc routine seq worker /* { dg-error "multiple loop axes" } */
void
f5a (void)
{
}

#pragma acc routine vector seq /* { dg-error "multiple loop axes" } */
void
f6 (void)
{
}

#pragma acc routine seq vector /* { dg-error "multiple loop axes" } */
void
f6a (void)
{
}

File: gcc/testsuite/gfortran.dg/goacc/routine-4.f90
! Test invalid calls to routines.

module param
  integer, parameter :: N = 32
end module param

program main
  use param
  integer :: i
  integer :: a(N)

  do i = 1, N
    a(i) = i
  end do

  !
  ! Seq routine tests.
  !

  !$acc parallel copy (a)
  !$acc loop
  do i = 1, N
     call seq (a)
  end do

  !$acc loop gang
  do i = 1, N
     call seq (a)
  end do

  !$acc loop worker
  do i = 1, N
     call seq (a)
  end do

  !$acc loop vector
  do i = 1, N
     call seq (a)
  end do
  !$acc end parallel

  !
  ! Gang routines loops.
  !

  !$acc parallel copy (a)
  !$acc loop
  do i = 1, N
     call gang (a)
  end do

  !$acc loop gang
  do i = 1, N
     call gang (a) ! { dg-error "incompatible parallelism with acc routine" }
  end do

  !$acc loop worker
  do i = 1, N
     call gang (a) ! { dg-error "incompatible parallelism with acc routine" }
  end do

  !$acc loop vector
  do i = 1, N
     call gang (a) ! { dg-error "incompatible parallelism with acc routine" }
  end do
  !$acc end parallel

  !
  ! Worker routines loops.
  !

  !$acc parallel copy (a)
  !$acc loop
  do i = 1, N
     call worker (a)
  end do

  !$acc loop gang
  do i = 1, N
     call worker (a)
  end do

  !$acc loop worker
  do i = 1, N
     call worker (a) ! { dg-error "incompatible parallelism with acc routine" }
  end do

  !$acc loop vector
  do i = 1, N
     call worker (a) ! { dg-error "incompatible parallelism with acc routine" }
  end do
  !$acc end parallel

  !
  ! Vector routines loops.
  !

  !$acc parallel copy (a)
  !$acc loop
  do i = 1, N
     call vector (a)
  end do

  !$acc loop gang
  do i = 1, N
     call vector (a)
  end do

  !$acc loop worker
  do i = 1, N
     call vector (a)
  end do

  !$acc loop vector
  do i = 1, N
     call vector (a) ! { dg-error "incompatible parallelism with acc routine" }
  end do
  !$acc end parallel
contains

  subroutine gang (a)
    !$acc routine gang
    integer, intent (inout) :: a(N)
    integer :: i

    do i = 1, N
       a(i) = a(i) - a(i)
    end do
  end subroutine gang

  subroutine worker (a)
    !$acc routine worker
    integer, intent (inout) :: a(N)
    integer :: i

    do i = 1, N
       a(i) = a(i) - a(i)
    end do
  end subroutine worker

  subroutine vector (a)
    !$acc routine vector
    integer, intent (inout) :: a(N)
    integer :: i

    do i = 1, N
       a(i) = a(i) - a(i)
    end do
  end subroutine vector

  subroutine seq (a)
    !$acc routine seq
    integer, intent (inout) :: a(N)
    integer :: i

    do i = 1, N
       a(i) = a(i) - a(i)
    end do
  end subroutine seq
end program main
